datafusion/sql/src/planner.rs

```diff
@@ -662,21 +662,33 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         ctes: &mut HashMap<String, LogicalPlan>,
         outer_query_schema: Option<&DFSchema>,
     ) -> Result<LogicalPlan> {
+        // From clause may exist CTEs, we should separate them from global CTEs.
+        // CTEs in from clause are allowed to be duplicated.
+        // Such as `select * from (WITH source AS (select 1 as e) SELECT * FROM source) t1, (WITH source AS (select 1 as e) SELECT * FROM source) t2;` which is valid.
+        // So always use original global CTEs to plan CTEs in from clause.
+        // Btw, don't need to add CTEs in from to global CTEs.
+        let origin_ctes = ctes.clone();
```

Comment:
the cloning here is unfortunate but I think unavoidable unless we moved to a stack based approach rather than a mutable hashtable.

Reply:
Yes, fortunately, it's not performance critical path.